code/game/objects/structures/reflector.dm

```diff
@@ -0,0 +1,177 @@
+/obj/structure/reflector
+	name = "reflector frame"
+	icon = 'icons/obj/stock_parts.dmi'
+	icon_state = "box_0"
+	desc = "A frame to create a reflector.\n<span class='notice'>Use <b>5</b> sheets of <b>glass</b> to create a 1 way reflector.\nUse <b>10</b> sheets of <b>reinforced glass</b> to create a 2 way reflector.\nUse <b>1 diamond</b> to create a reflector cube.</span>"
+	anchored = 0
+	density = 1
+	layer = 3
+	var/finished = 0
+
+
+/obj/structure/reflector/bullet_act(obj/item/projectile/P)
+	var/turf/reflector_turf = get_turf(src)
+	var/turf/reflect_turf
+	var/new_dir = get_reflection(src.dir,P.dir)
+	if(!istype(P, /obj/item/projectile/beam))
+		return..()
+	if(new_dir)
+		reflect_turf = get_step(reflect_turf, new_dir)
+	else
+		visible_message("<span class='notice'>[src] is hit by the [P]!</span>")
+		new_dir = 0
+		return ..() //Hits as normal, explodes or emps or whatever
+
+	visible_message("<span class='notice'>[P] bounces off of the [src]</span>")
+	reflect_turf = get_step(loc,new_dir)
+
+	P.original = reflect_turf
+	P.starting = reflector_turf
+	P.current = reflector_turf
+	P.yo = reflect_turf.y - reflector_turf.y
+	P.xo = reflect_turf.x - reflector_turf.x
+	new_dir = 0
+	return - 1
+
+
+/obj/structure/reflector/attackby(obj/item/W, mob/user, params)

```

Comment:
The below comments apply to so many lines in this proc I wasn't going to write it out for each. If you already have a `mob/user` parameter then use it instead of `usr`. Also check for multiple instances of implied `src` and `\the`, `the` etc.